feat: partial compilation of directives

JoostK opened the pull request:
See individual commits.

JoostK committed:
refactor(compiler-cli): extend AST object and value helpers

This introduces `AstObject.toMap` as an alternative to `AstObject
.toLiteral`, and adds `AstValue.getSymbolName` to query the symbol name
of a value using the encapsulated AST host.

JoostK committed:
test(compiler-cli): run the linker transform in the compliance test's prelink target

In PR #38938 an additional Bazel target was introduced for the compliance
tests, as preparation to run the compliance tests in partial compilation
mode and then apply the linker transform. The linker plugin itself was
not available at the time but has since been implemented, so this commit
updates the prelink target of the compliance tests to apply the linker
transform using the Babel plugin.

Actually emitting partial compilations to be transformed will be done in
follow-up work.

JoostK committed:
feat(compiler-cli): partial compilation of directives

This commit implements partial code generation for directives, which
will be transformed by the linker plugin to fully AOT compiled code in
follow-up work.

JoostK committed:
feat(compiler-cli): implement partial directive declaration linking

This commit implements the logic to compile a partial declaration of a
directive into its full AOT compilation output.

JoostK committed:
fixup! refactor(compiler-cli): extend AST object and value helpers

petebacondarwin commented on packages/compiler-cli/test/compliance/prelink/bootstrap.ts:
```diff
@@ -16,11 +19,43 @@ import {CompileFn, doCompile, setCompileFn} from '../mock_compile';
+ *
+ * @param file The file name and its source to be transformed using the linker.
+ * @param linkerPlugin The linker plugin to apply.
+ * @returns The file's source content which has been transformed using the linker.
```
```suggestion
@returns The file's source content, which has been transformed using the linker if necessary.
```

petebacondarwin commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+function compileQuery(query: R3QueryMetadata): o.LiteralMapExpr {
+  const meta = new DefinitionMap();
+  meta.set('propertyName', o.literal(query.propertyName));
+  meta.set('first', o.literal(query.first));
+  meta.set(
+      'predicate', Array.isArray(query.predicate) ? asLiteral(query.predicate) : query.predicate);
+  meta.set('descendants', o.literal(query.descendants));
+  meta.set('read', query.read);
+  meta.set('static', o.literal(query.static));
```
All of these except for `predicate` are either optional already (e.g. `read`) or are simple booleans that can be made optional with suitable defaults, yes?

JoostK replied:
Correct. Done.

petebacondarwin commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+  definitionMap.set('inputs', conditionallyCreateMapObjectLiteral(meta.inputs, true));
+  definitionMap.set('outputs', conditionallyCreateMapObjectLiteral(meta.outputs));
+
+  definitionMap.set('host', compileHostMetadata(meta.host));
```
If the result of `compileHostMetadata()` is null we should not add this property.

JoostK replied:
It won't as mentioned before.

petebacondarwin commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+ */
+function compileHostMetadata(meta: R3HostMetadata): o.LiteralMapExpr|null {
+  const hostMetadata = new DefinitionMap();
+  const attributes = o.literalMap(Object.keys(meta.attributes).map(key => {
```
The code to add `attributes`, `listeners` and `properties` is very similar. Could we DRY it up by using a helper like:

```ts
function addMapToDefinition(
    definition: DefinitionMap, propertyName: string,
    map: {[key: string]: string|o.Expression}): void {
  const keys = Object.keys(map);
  if (keys.length > 0) {
    const literalMap = o.literalMap(Object.keys(map).map(key => {
      const rawValue = map[key];
      const value = typeof rawValue === 'string' ? o.literal(rawValue) : rawValue
      return {key, value, quoted: true};
    }));
    definition.set(propertyName, literalMap);
  }
}
```

petebacondarwin replied:
Also I wonder whether we always need to quote the key?
I see in `mapToExpression()`, there is:

```ts
quoted: UNSAFE_OBJECT_KEY_NAME_REGEXP.test(minifiedName),
```

JoostK replied:
`mapToExpression` is careful about only using quotes when needed, as non-quoted keys will be subject to Closure's minification which is desirable here, to correspond with the minification done to the respective class properties. We don't need that here so I choose to always use quotes.

I did refactor the logic of transforming an object literal into an object literal expression a bit, but a bit differently to what you suggested.

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -5,21 +5,145 @@
       metaObj: AstObject<TExpression>): o.Expression {
-    throw new Error('Not implemented.');
+    const meta = toR3DirectiveMeta(metaObj, code, sourceUrl);
+    const def = compileDirectiveFromMetadata(meta, constantPool, makeBindingParser());
```
Normally `makeBindingParser()` takes the interpolation config, but I guess since directives have no template there is not need to specify interpolation markers, right?
But we should consider this when implementing the component declaration linking.

JoostK replied:
Good point, yes indeed this becomes relevant for components.

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -5,21 +5,145 @@
+ */
+export function toR3DirectiveMeta<TExpression>(
+    metaObj: AstObject<TExpression>, code: string, sourceUrl: string): R3DirectiveMetadata {
+  const typeName = metaObj.getValue('type').getSymbolName() ?? 'anonymous';
```
Is "anonymous" a valid scenario? Perhaps we should throw an error here?
From digging through the code, it looks like this directive type name is used to create names for queries and template functions. I think it would be worth, if not erroring, making sure that this is unique for the given constant pool?

petebacondarwin replied:
The `type` property is not optional on the `R3DeclareDirectiveMetadata` interface, so it seems unlikely (impossible?) that a directive would be defined as an anonymous class?

JoostK replied:
Yes I agree this case should throw instead of the magic anonymous name.

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -5,21 +5,145 @@
+export function toR3DirectiveMeta<TExpression>(
+    metaObj: AstObject<TExpression>, code: string, sourceUrl: string): R3DirectiveMetadata {
+  const typeName = metaObj.getValue('type').getSymbolName() ?? 'anonymous';
+  const range = metaObj.getValue('type').getRange();
```
BIKESHED: `typeRange`.

JoostK replied:
Instead of painting the bikeshed I tore it down; inlined this variable into its only usage. 🚲🏠

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -5,21 +5,145 @@
+
+  const values = value.getArray().map(innerValue => innerValue.getString());
+  if (values.length !== 2) {
+    throw new FatalLinkerError(value.expression, 'Unsupported input, expected exactly two strings');
```
Might be worth mentioning the other available option here:

```
'Unsupported input property, expected a string or an array containing exactly two strings'
```

JoostK replied:
Done

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -5,21 +5,145 @@
+}
+
+export function createSourceSpan(range: Range, code: string, sourceUrl: string): ParseSourceSpan {
+  const sourceFile = new ParseSourceFile(code, sourceUrl);
```
We will end up creating lots of identical `ParseSourceFile` objects - one for each partial in the source file. I wonder if we ought to have a cache of these somehow? In fact, the `PartialDirectiveLinkerVersion1` would be a suitable place to hold this cache, since its life is the same as the plugin.

JoostK replied:
I don't follow. All these calls would be unique per type, no?

petebacondarwin replied:
My suggestion is that we have a map of `sourceUrl` to `ParseSourceFile` stored somewhere. There could be a map per partial linker (or better per `PartialLinkerSelector` - or perhaps even passed through to that from the `FileLinker`?)

Then this function would either be moved inside the partial linker of have the map passed to it.

JoostK replied:
Oh, I was confusing this to be about the `ParseSourceSpan`, but it's just the file. I don't think it's a concern that we need to have a cache for these; they are effectively free and will be GC'd after the metadata has been compiled (practically immediately). I'd rather leave as-is.

petebacondarwin replied:
Yeah - fair enough

petebacondarwin commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_linker_selector.ts:
```diff
@@ -10,7 +10,7 @@ import {PartialDirectiveLinkerVersion1} from './partial_directive_linker_1';
 import {PartialLinker} from './partial_linker';
 
 export class PartialLinkerSelector<TStatement, TExpression> {
```
You can remove the `TStatement` here too.

JoostK replied:
Done.

petebacondarwin reviewed (changes requested):
Nice! I love how clean the whole PR feels. It is really easy to follow along with what it does.

petebacondarwin commented:
Also, I am not sure whether these commits should be `feat` or `refactor`... In any case, I think that the 3rd commit message should probably reworded slightly:

```
feat(compiler-cli): implement partial directive declaration linking 

This commit implements the logic to compile a partial declaration of a
directive into its full AOT compilation output.
```

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -171,6 +171,26 @@ export class DirectiveDecoratorHandler implements
+        type: res.type,
+      }
+    ];
+  }
```
Works for me, done!

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -171,6 +171,26 @@ export class DirectiveDecoratorHandler implements
 
+  compilePartial(
+      node: ClassDeclaration, analysis: Readonly<DirectiveHandlerData>,
+      resolution: Readonly<unknown>): CompileResult|CompileResult[] {
```
Switched to `CompileResult[]`. The interface declares the union type, so my IDE copied that when implementing the method.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -0,0 +1,139 @@
+  // Version number of the metadata format. This is used to evolve the metadata
+  // interface later - the linker will be able to detect which version a library
+  // is using and interpret its metadata accordingly.
+  version: 1;
```
Yes, but doing that now would also require changes to the selector logic etc. which I felt was better left for another PR.

petebacondarwin replied:
OK

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -0,0 +1,139 @@
+  // Reference to the directive class itself.
+  type: o.Expression;
+
+  // Map of inputs, keyed by the name of the input field.
```
I'm surprised these descriptions are different in the first place; I copied this interface from the spike code, but I'm not sure how that one was put together. Tweaked the text a bit, also in `R3DirectiveMetadata`, to hopefully make it more clear.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -0,0 +1,139 @@
+  /**
+   * Whether the component has an inheritance clause.
+   */
+  usesInheritance: boolean;
```
I wasn't sure about these and left all booleans as required. I'm fine with changing them to a sane default, though.

petebacondarwin replied:
Actually, I think it is quite relevant. Some libraries (e.g. Material) might contain many many directives and components. While the final application production code will not contain this, there is a cost both in space and parse time, etc for the published library code.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -0,0 +1,139 @@
+                             // likely be dropped
+  usesOnChanges: boolean;
+
+  // A reference to the `@angular/core` ES module, which allows access
+  // to all Angular exports, including Ivy instructions.
```
Changed all of these.

JoostK commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+ * Gathers the declaration fields for a directive into a `DefinitionMap`. This allows for reusing
+ * this logic for components, as they extend the directive metadata.
+ */
+export function baseDirectivePartialCompile(meta: R3DirectiveMetadata): DefinitionMap {
```
This was mirrored from the full AOT compiler. I like `createDirectiveDefinitionMap`, going with that.

JoostK commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+export function baseDirectivePartialCompile(meta: R3DirectiveMetadata): DefinitionMap {
+  const definitionMap = new DefinitionMap();
+
+  definitionMap.set('version', o.literal(1));
```
Leaving for future PR.

JoostK commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+    definitionMap.set('selector', o.literal(meta.selector));
+  }
+
+  definitionMap.set('inputs', conditionallyCreateMapObjectLiteral(meta.inputs, true));
+  definitionMap.set('outputs', conditionallyCreateMapObjectLiteral(meta.outputs));
+
+  definitionMap.set('host', compileHostMetadata(meta.host));
+
+  definitionMap.set('providers', meta.providers);
```
Actually `DefinitionMap` has a trick up its sleeve:

https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/compiler/src/render3/view/util.ts#L151

so passing in null here does in fact result in not having that property.

JoostK commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -0,0 +1,131 @@
+    definitionMap.set('exportAs', asLiteral(meta.exportAs));
+  }
+
+  definitionMap.set('usesInheritance', o.literal(meta.usesInheritance));
+  definitionMap.set('fullInheritance', o.literal(meta.fullInheritance));
+  definitionMap.set('usesOnChanges', o.literal(meta.lifecycle.usesOnChanges));
```
Done.

JoostK committed:
fixup! test(compiler-cli): run the linker transform in the compliance test's prelink target

JoostK committed:
fixup! feat(compiler-cli): partial compilation of directives

JoostK committed:
fixup! feat(compiler-cli): implement partial directive declaration linking

JoostK committed:
fixup! feat(compiler-cli): partial compilation of directives

JoostK committed:
fixup! feat(compiler-cli): partial compilation of directives

petebacondarwin reviewed (approved):
LGTM!

JoostK commented on packages/compiler/src/render3/partial/directive.ts:
```diff
@@ -77,12 +80,18 @@ export function baseDirectivePartialCompile(meta: R3DirectiveMetadata): Definiti
   meta.set(
       'predicate', Array.isArray(query.predicate) ? asLiteral(query.predicate) : query.predicate);
-  meta.set('descendants', o.literal(query.descendants));
+  if (query.descendants) {
```
I would guess that `ViewChild` is most common, but honestly I have no idea what the distribution actually is.

I did consider basically this question but then for the `first` flag, which I suspect is more often `true` than `false`. But it did indeed feel simpler/more consistent to go with `false` as a default here.

jessicajaniuk reviewed (approved):
This looks really clean to me! LGTM!

petebacondarwin reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented on goldens/public-api/core/core.d.ts:
```diff
@@ -1,3 +1,6 @@
+/** @codeGenApi */
+export declare function $ngDeclareDirective(decl: unknown): unknown;
```
Just curious if we can use `ɵɵ` here (i.e. `ɵɵngDeclareDirective`)?

JoostK replied:
The `$ng` dollar syntax is as originally proposed by @alxhub. I suppose it's to make the linker more rigid in detecting these calls at link-time.

AndrewKushnir replied:
Thanks for additional information @JoostK!

I'd vote for a consistent way to annotate such APIs with `ɵ`s :) If there are advantages of using a different prefix - may be we can document it in the code somewhere (in the `ngDeclareDirective ` function docs) so that we can refer to it?

JoostK replied:
Your vote has been recorded and will be included in the final count. Due to the large number of votes by mail during this election, it might take a while to count them all.

Just kidding; I had a chat with Alex and we did end up changing the function to use `ɵɵ` as prefix.

AndrewKushnir replied:
Thanks @JoostK 😄

JoostK committed:
refactor(compiler-cli): rename `$ngDeclareDirective`/`$ngDeclareComponent` to use `ɵɵ` prefix

For consistency with other generated code, the partial declaration
functions are renamed to use the `ɵɵ` prefix which indicates that it is
generated API.

This commit also removes the declaration from the public API golden
file, as it's not yet considered stable at this point. Once the linker
is finalized will these declaration function be included into the golden
file.

AndrewKushnir commented:
After changing `$ng` prefix to `ɵɵ` one it looks like `public-api` group approval is not needed for this PR.

AndrewKushnir commented:
[Presubmit](http://test/OCL:340562894:BASE:340562903:1604456308358:2278dbc).

AndrewKushnir commented:
FYI, presubmit is successful, I'm removing the "presubmit" label.
